### ISSUE TITLE
code error

### ISSUE THREAD

**Issue by matrxcrystal** [ExistentialAudio/BlackHole#678]
### Please Confirm

- [X] I have read the **[FAQ](https://github.com/ExistentialAudio/BlackHole#faq) and [Wiki](https://github.com/ExistentialAudio/BlackHole/wiki)** where most common issues can be resolved
- [X] I have searched **[Discussions](https://github.com/ExistentialAudio/BlackHole/discussions)** to see if the same question has already been asked
- [X] This is a **bug** and not a question about audio routing or configuration, which should be posted in [Discussions](https://github.com/ExistentialAudio/BlackHole/discussions)

### macOS Version

macOS 10.15 Catalina

### BlackHole Build(s) Affected

- [X] 2 channel
- [X] 16 channel
- [X] 64 channel
- [X] other/custom build

### Describe the bug

I checked blackHole code and found several serious problems
1. the code from apple simple code `NullAudio`
`((AudioObjectID*)outData)[0] = kObjectID_Box;
((AudioObjectID*)outData)[0] = kObjectID_Device;`
change to
`((AudioObjectID*)outData)[0] = kObjectID_Box;
((AudioObjectID*)outData)[1] = kObjectID_Device;
((AudioObjectID*)outData)[2] = kObjectID_Device;`
2. volume control code error
in proprety kAudioLevelControlPropertyScalarValue and kAudioLevelControlPropertyDecibelValue
using  db = 20 * lg (p1/p2)  
but in proprety kAudioLevelControlPropertyConvertScalarToDecibels and kAudioLevelControlPropertyConvertDecibelsToScalar
using square/sqrtf
3. one param gVolume_Master_Value can‘t control input and output in BlackHole_DoIOOperation
4. in BlackHole_DoIOOperation method read and write gRingBuffer , but not use io mutex lock

### Reproduction Steps

none

### Expected Behavior

none

**Comment by devinroth** [comment]
1. Noted.
2. Noted.
3. That’s by design. I used to have them separate. What issue do you have here?
4. Also by design. One should never use mutex on a real-time audio thread. I’ve spent lots of time thinking about this and running various tests. It’s kind of a lose lose situation. Using mutex will guarantee no race conditions but will lock the thread potentially causing audio errors on other threads. On the other hand, no mutex will guarantee the threads will return in time but potentially have an issue when writing to the same buffer location at the same time. There is no risk of a crash. After numerous tests I’ve chosen the latter and guess what, I’ve never had any issues with it.


Devin

**Comment by devinroth** [comment]
Input and output should never access the same buffer location at the same time. If they do, it should be an error no matter what. A lock won’t prevent and error and potentially cause more.  And writing to the buffer only ever happens with a single thread. HAL already mixed the audio. It’s better to not lock up the real-time thread.

I know it goes against everything we’re taught but there’s “what we should do” and there’s “the right thing to do”.

Devin

**Comment by matrxcrystal** [comment]
3. I don't have any problems, I just want to know why the BlackHole code and apple simple code are not consistent. It seems to me that you should use gVolume_Output_Master_Value in the kAudioServerPlugInIOOperationReadInput callback to control the output volume and gVolume_Input_Master_Value in the kAudioServerPlugInIOOperationWriteMix callback to control the input volume. In this way, I can better make the driver work in the system sound settings interface
4. I saw in the apple simple demo (c++) that a mutex is added to the IO operation. I very much agree with your point of view, the static ringbuffer will not be released, obviously this will not cause a crash. If the input and output will not access the ringbuffer at the same time, I think it is very reasonable to not lock. Currently I need to do some tests to confirm this.

**Comment by devinroth** [comment]
3. That’s how it was originally but after too many people complaining about volume not working because they didn’t realize the input and output have separate control, I disabled the volume on the input. Unfortunately that caused a whole number of different complaints because volume on the input no longer worked. That’s when I got the idea of merging them. Nobody has complained since.

4. One of the rules of real-time audio threads is don’t do anything that will lock the thread. They inevitably cause issues. That means don’t lock, don’t use ObjC/Swift, and don’t allocate memory. Unfortunately Apple doesn’t always follow these guidelines and build audio apps that have issues.

Devin
